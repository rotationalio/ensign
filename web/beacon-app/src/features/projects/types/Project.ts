import { MemberResponse } from '@/features/members/types/memberServices';
export type Project = {
  created: string;
  id: string;
  modified: string;
  name: string;
  tenant_id: string;
  description?: string;
  status?: string;
  owner: Partial<MemberResponse>;
};
