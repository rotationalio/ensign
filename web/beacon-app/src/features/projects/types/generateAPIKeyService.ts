export interface NewApiKey {
  name: string;
  permissions: string[];
}
