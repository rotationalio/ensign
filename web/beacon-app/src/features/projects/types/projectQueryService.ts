import { UseMutateFunction } from '@tanstack/react-query';

export interface ProjectQueryMutation {
  getProjectQuery: UseMutateFunction<ProjectQueryDTO, unknown, ProjectQueryDTO, unknown>;
  reset(): void;
  projectQuery: any;
  hasProjectQueryFailed: boolean;
  wasProjectQueryCreated: boolean;
  isCreatingProjectQuery: boolean;
  error: any;
}

export interface ProjectQueryDTO {
  projectID: string;
  query: string;
}
