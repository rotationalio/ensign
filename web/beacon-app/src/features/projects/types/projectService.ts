export interface ProjectsResponse {
  project: ProjectResponse[];
  prev_page_token: string;
  next_page_token: string;
}

export type ProjectResponse = {
  id: string;
  name: string;
  created: string;
  modified?: string;
  description?: string;
  status?: string;
};

export type ProjectDetailQuery = {
  // getProject: (id: ProjectDetailDTO) => Promise<ProjectResponse | undefined>;
  project: any;
  hasProjectFailed: boolean;
  wasProjectFetched: boolean;
  isFetchingProject: boolean;
  error: any;
};

export type ProjectDetailDTO = Pick<ProjectResponse, 'id'>;

export type ProjectsQuery = {
  getProjects(): void;
  projects: any;
  hasProjectsFailed: boolean;
  wasProjectsFetched: boolean;
  isFetchingProjects: boolean;
  error: any;
};

export interface ProjectQuickViewData {
  name: string;
  value: number;
  units?: string;
  percent?: number;
}

export interface ProjectQuickViewResponse {
  data: ProjectQuickViewData[];
}

export interface ProjectQuickViewQuery {
  getProjectQuickView: () => void;
  hasProjectQuickViewFailed: boolean;
  isFetchingProjectQuickView: boolean;
  projectQuickView: any;
  wasProjectQuickViewFetched: boolean;
  error: any;
}
