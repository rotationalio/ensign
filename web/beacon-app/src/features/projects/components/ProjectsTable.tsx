import { t } from '@lingui/macro';
import { Table } from '@rotational/beacon-core';
import { ErrorBoundary } from '@sentry/react';
import React, { useCallback, useMemo, useState } from 'react';
import toast from 'react-hot-toast';
import { useNavigate } from 'react-router-dom';

import { PATH_DASHBOARD } from '@/application/routes/paths';
import { formatDate } from '@/utils/formatDate';

import { RenameProjectModal } from '../components/RenameProject';
import { Project } from '../types/Project';

type ProjectTableProps = {
  projects: Project[];
  isLoading?: boolean;
};

const ProjectsTable: React.FC<ProjectTableProps> = ({ projects, isLoading = false }) => {
  const navigate = useNavigate();
  const initialColumns = useMemo(
    () => [
      { Header: t`Project ID`, accessor: 'id' },
      { Header: t`Project Name`, accessor: 'name' },
      {
        Header: t`Description`,
        accessor: (p: Project) => {
          const description = p?.description;
          if (!description) {
            return '---';
          }
          // cut off description at 100 characters
          return description?.length > 100
            ? `${description?.slice(0, 100)}...`
            : description || '---';
        },
      },
      {
        Header: 'Status',
        accessor: (p: Project) => {
          const status = p?.status;
          return status || '---';
        },
      },
      {
        Header: t`Date Created`,
        accessor: (date: any) => {
          return formatDate(new Date(date?.created));
        },
      },
      { Header: 'Actions', accessor: 'actions' },
    ],
    []
  ) as any;

  const initialState = { hiddenColumns: ['id'] };

  const [openRenameProjectModal, setOpenRenameProjectModal] = useState<{
    open: boolean;
    project: Project | null;
  }>({
    open: false,
    project: null,
  });

  const handleRenameProjectClick = (project: Project) => {
    setOpenRenameProjectModal({ open: true, project });
  };

  const handleChangeOwnerClick = (_projectId: string) => {};

  const getProjects = useCallback((projects: Project[]) => {
    return (projects || []).map((project: Project) => ({
      ...project,
      actions: [
        { label: 'Rename project', onClick: () => handleRenameProjectClick(project) },
        { label: 'Change owner', onClick: () => handleChangeOwnerClick(project?.id) },
      ],
    }));
  }, []);

  const handleModalClose = () =>
    setOpenRenameProjectModal({ ...openRenameProjectModal, open: false });

  const handleRedirection = (row: any) => {
    if (!row?.values?.id) {
      toast.error(
        t`Sorry, we are having trouble redirecting you to your project. Please try again.`
      );
    }
    navigate(`${PATH_DASHBOARD.PROJECTS}/${row?.values?.id}`);
  };

  return (
    <div className="mx-4">
      <ErrorBoundary
        fallback={
          <div className="item-center my-auto flex w-full text-center font-bold text-danger-500">
            <p>
              Sorry we are having trouble listing your projects, please refresh the page and try
              again.
            </p>
          </div>
        }
      >
        <RenameProjectModal {...openRenameProjectModal} handleModalClose={handleModalClose} />
        <Table
          trClassName="text-sm hover:bg-gray-100"
          columns={initialColumns}
          initialState={initialState}
          data={getProjects(projects) || []}
          onRowClick={(row: any) => {
            handleRedirection(row);
          }}
          isLoading={isLoading}
        />
      </ErrorBoundary>
    </div>
  );
};

export default ProjectsTable;
