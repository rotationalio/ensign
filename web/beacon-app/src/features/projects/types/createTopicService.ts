import { UseMutateFunction } from '@tanstack/react-query';

import { Topic } from '@/features/topics/types/topicService';

export interface TopicMutation {
  createTopic: UseMutateFunction<Topic, unknown, NewTopicDTO, unknown>;
  reset(): void;
  topic: any;
  hasTopicFailed: boolean;
  wasTopicCreated: boolean;
  isCreatingTopic: boolean;
  error: any;
}

export interface NewTopic {
  topic_name: string;
}

export type NewTopicDTO = {
  projectID: string;
} & NewTopic;

export const isTopicCreated = (mutation: TopicMutation): mutation is Required<TopicMutation> =>
  mutation.wasTopicCreated && mutation.topic != undefined;
