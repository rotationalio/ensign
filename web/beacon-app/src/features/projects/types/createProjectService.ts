import { UseMutateFunction } from '@tanstack/react-query';

import { ProjectResponse } from './projectService';

export interface ProjectMutation {
  createNewProject: UseMutateFunction<ProjectResponse, unknown, NewProjectDTO, unknown>;
  reset(): void;
  project: ProjectResponse;
  hasProjectFailed: boolean;
  wasProjectCreated: boolean;
  isCreatingProject: boolean;
  error: any;
}

export type NewProjectDTO = Pick<ProjectResponse, 'name' | 'description'> & {
  tenantID: string;
};

export const isProjectCreated = (
  mutation: ProjectMutation
): mutation is Required<ProjectMutation> =>
  mutation.wasProjectCreated && mutation.project != undefined;
