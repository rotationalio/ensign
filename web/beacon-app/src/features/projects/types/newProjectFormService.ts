/* eslint-disable prettier/prettier */
import { t } from '@lingui/macro';
import { useFormik } from 'formik';
import { object, string } from 'yup';

import { NewProjectDTO } from './createProjectService';

export const FORM_INITIAL_VALUES = {
  name: '',
  description: '',
} satisfies Omit<NewProjectDTO, 'tenantID'>;

export const FORM_VALIDATION_SCHEMA = object({
  name: string()
    .trim()
    .required(t`Project name is required.`)
    .max(512, t`Project name must be less than 512 characters.`),
  description: string().notRequired(),
});
export const FORM_OPTIONS = (onSubmit: any) => ({
  initialValues: FORM_INITIAL_VALUES,
  validationSchema: FORM_VALIDATION_SCHEMA,
  onSubmit,
});

export const useNewProjectForm = (onSubmit: any) => useFormik(FORM_OPTIONS(onSubmit));
