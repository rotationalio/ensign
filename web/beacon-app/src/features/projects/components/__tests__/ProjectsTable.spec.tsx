import { render } from '@testing-library/react';
import jest from 'jest-mock';
import React from 'react';
import { vi } from 'vitest';

//import { customRender } from '../../../../utils/test-utils';
import { Project } from '../../types/Project';
import ProjectsTable from '../ProjectsTable';

const renderComponent = (props) => {
  return render(<ProjectsTable {...props} />);
};
vi.mock('react-router-dom', async (importOrginial: any) => ({
  ...importOrginial,
  useNavigate: () => jest.fn(),
}));

vi.mock('@lingui/macro', async (importOrginial: any) => ({
  ...importOrginial,
  t: () => jest.fn(),
  Trans: () => jest.fn(),
}));
describe('ProjectsTable', () => {
  // mock useNavigate hook

  //   vi.mock('react-table', async (importOrginial: any) => ({
  //     ...importOrginial,
  //     useTable: () => jest.fn(),
  //     useFilters: () => jest.fn(),
  //     useSortBy: () => jest.fn(),
  //     usePagination: () => jest.fn(),
  //     Row: () => jest.fn(),

  //     Column: () => jest.fn(),
  //   }));

  const MockProjectProps = {
    projects: [
      {
        id: '1',
        name: 'Project 1',
        tenant_id: '1',
        modified: '2021-01-01',
        description: 'Project 1 description',
        status: 'Active',
        created: '2021-01-01',
      },
      {
        id: '2',
        name: 'Project 2',
        description: 'Project 2 description',
        status: 'Inactive',
        created: '2021-01-02',
      },
    ] as Project[],
  };

  it('should render table with correct columns', () => {
    const { container } = renderComponent(MockProjectProps);

    expect(container).toMatchSnapshot();
    // const table = screen.getByRole('table');
    // expect(table).toBeInTheDocument();
    // expect(screen.getByText('Project ID')).toBeInTheDocument();
    // expect(screen.getByText('Project Name')).toBeInTheDocument();
    // expect(screen.getByText('Description')).toBeInTheDocument();
    // expect(screen.getByText('Status')).toBeInTheDocument();
    // expect(screen.getByText('Date Created')).toBeInTheDocument();
  });
});
