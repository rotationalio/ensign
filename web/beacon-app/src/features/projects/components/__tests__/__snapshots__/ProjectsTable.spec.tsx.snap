// Vitest Snapshot v1

exports[`ProjectsTable > should render table with correct columns 1`] = `
<div>
  <div
    class="mx-4"
  >
    <div
      class="item-center my-auto flex w-full text-center font-bold text-danger-500"
    >
      <p>
        Sorry we are having trouble listing your projects, please refresh the page and try again.
      </p>
    </div>
  </div>
</div>
`;
